framework/include/base/Assembly.h

```diff
@@ -672,6 +672,43 @@ class Assembly
 
   /// Temporary work vector to keep from reallocating it
   std::vector<dof_id_type> _temp_dof_indices;
+
+public:
+  /**
+   * We can't add NodalBC values to the Jacobian (or preconditioning)
+   * matrix at the time they are computed -- we instead need to
+   * overwrite an entire row of values with the Jacobian associated to
+   * the NodalBC, which may be coupled to other variables, *after*
+   * matrix assembly is finished.  Hence, the Assembly object
+   * maintains a reference to this data which NodalBCs can write to,
+   * and which can be used later to properly set NodalBC jacobians.
+   * We use numeric_index_type for the index arrays (rather than
+   * dof_id_type) since that is what the SparseMatrix interface uses,
+   * but at the time of this writing, those two types are equivalent.
+   */
+  class CachedNodalBCData
+  {
+  public:
+    // Clears the cached data arrays
+    void clear()
+    {
+      cached_nodal_bc_vals.clear();
+      cached_nodal_bc_rows.clear();
+      cached_nodal_bc_cols.clear();
+    }
+
+    std::vector<Real> cached_nodal_bc_vals;
+    std::vector<numeric_index_type> cached_nodal_bc_rows;
+    std::vector<numeric_index_type> cached_nodal_bc_cols;
+  };
+
+  /**
+   * Return a writeable reference to the cached Nodal BC data.
+   */
+  CachedNodalBCData & cachedNodalBCData() { return _cached_nodal_bc_data; }
```

Comment:
I don't really understand the point of this - if it's just going to return a writeable reference - why not just make the member variable public?

I mean: I'm not 100% sure that's the right idea anyway - but this seems to just straddle the line...